In [15]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [16]:
# Load the given datasets
numerical = pd.read_csv("chronic_kidney_disease_numerical.csv")
categorical = pd.read_csv("chronic_kidney_disease_categorical.csv")

# Print the data
print(numerical)
print(categorical)

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2     5500.0   5.7  
3     8200.0   4.5  
4     

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: 
The dataset is essentially hospital-provided data (over the span of 2 months) from which insight can be garnered in regards to chronic kidney disease prediction. The features are the various types of health data, such as blood pressure, age, red blood cells, and hypertension. The primary label is whether or not an individual has chronic kidney disease (ckd and notckd). The features in the datasets are numerical and categorical. Age, blood pressure, and blood urea are examples of numerical columns. Sugar, bacteria, and specific gravity are examples of categorical columns.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [17]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
duplicate_rows_numerical = numerical[numerical.duplicated(subset=['unique_id'], keep=False)]

if duplicate_rows_numerical.empty:
    print("No duplicate rows in the numerical dataset\n")
else:
    print(f"Total number of duplicate rows in the numerical dataset: {duplicate_rows_numerical.shape[0]}\n")
    print("Duplicate numerical rows:\n")
    print(duplicate_rows_numerical)

# Drop these duplicate rows
numerical = numerical.drop_duplicates(subset=['unique_id'])

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
duplicate_rows_categorical = categorical[categorical.duplicated(subset=['unique_id'], keep=False)]

if duplicate_rows_categorical.empty:
    print("No duplicate rows in the categorical dataset\n")
else:
    print(f"Total number of duplicate rows in the categorical dataset: {duplicate_rows_categorical.shape[0]}\n")
    print("Duplicate categorical rows:\n")
    print(duplicate_rows_categorical)

categorical = categorical.drop_duplicates(subset=['unique_id'])

Total number of duplicate rows in the numerical dataset: 59

Duplicate numerical rows:

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
15      224481  54.0   60.0  125.0   21.0  1.3  137.0  3.4  15.0  46.0   
35      431871  57.0   60.0  132.0   18.0  1.1  150.0  4.7  15.4  42.0   
42      441170  79.0   80.0  111.0   44.0  1.2  146.0  3.6  16.3  40.0   
53      261137  35.0   60.0  105.0   39.0  0.5  135.0  3.9  14.7  43.0   
54      721270  58.0   70.0  102.0   48.0  1.2  139.0  4.3  15.0  40.0   
61      872034  70.0   90.0  184.0   98.6  3.3  138.0  3.9   5.8   NaN   
78      501321   4.0    NaN   99.0   23.0  0.6  138.0  4.4  12.0  34.0   
79      283427  55.0   80.0  140.0   49.0  0.5  150.0  4.9  15.7  47.0   
81      302906  59.0  100.0    NaN   96.0  6.4    NaN  NaN   6.6   NaN   
98      708255   3.0    

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer:

In [18]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined_dataset = pd.merge(numerical, categorical, on='unique_id', how='outer')

#Print the combined dataset
print("Combined dataset:\n")
print(combined_dataset)

Combined dataset:

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       103570  24.0  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0  ...   
2       104675  65.0  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0  ...   
3       105923  90.0  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0  ...   
4       107297  60.0  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   
397     996412  70.0  60.0  120.0   40.0   0.5  140.0  4.6  16.0  43.0  ...   
398     997629  35.0   NaN   99.0   30.0   0.5  135.0  4.9  15.4  48.0  ...   
399     997876  64.0  70.0  107.0   15.0   NaN    NaN  NaN  12.8  38.0  ...   

           pc         pcc       

## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [19]:
# Calculate the percentage of rows that contain atleast one missing value
missing_rows = combined_dataset.isnull().any(axis=1).sum()
total_rows = combined_dataset.shape[0]
percentage_missing = (missing_rows / total_rows) * 100

# Print %
print(f"Percentage of rows containing at least one missing value: {percentage_missing:.2f}%\n")

# Drop these rows from the dataset
combined_no_missing = combined_dataset.dropna()

# Print the Dataset
print("Combined dataset without rows containing missing values):\n")
print(combined_no_missing)

Percentage of rows containing at least one missing value: 61.75%

Combined dataset without rows containing missing values):

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
6       109053  64.0  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  ...   
7       114717  41.0  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  ...   
8       118191  42.0  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  ...   
9       118596  74.0  60.0   88.0   50.0   0.6  147.0  3.7  17.2  53.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
390     980291  62.0  70.0  122.0   42.0   1.7  136.0  4.7  12.6  39.0  ...   
391     991031  30.0  80.0   89.0   42.0   0.5  139.0  5.0  16.7  52.0  ...   
394     994377  71.0  70.0   81.0   18.0   0.8  145.0  5.0  14.7  44.0  ...   
395     995177  69.0  70.0  214.0   96.0   6.3  120.0  3.9   9.4  28.0  ...   
396   

## Part 3.4: Sort the dataset according to the Labels

In [20]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
sorted_dataset = combined_no_missing.sort_values(by='Target')
sorted_dataset.reset_index(drop=True, inplace=True)

# Print the dataset
print("Sorted dataset according to 'Target' column:\n")
print(sorted_dataset)

Sorted dataset according to 'Target' column:



     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

           pc         pcc          ba  htn   dm  c

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [21]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
encoded_dataset = pd.get_dummies(sorted_dataset, columns=categorical_columns, drop_first=True)

# Print the dataset
print("Dataset after encoding categorical data:\n")
print(encoded_dataset)

Dataset after encoding categorical data:

     unique_id   age    bp    bgr     bu    sc    sod  pot  hemo   pcv  ...  \
0       102677  59.0  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  ...   
1       337834  48.0  70.0  117.0   56.0   3.8  111.0  2.5  11.2  32.0  ...   
2       343710  61.0  80.0  173.0  148.0   3.9  135.0  5.2   7.7  24.0  ...   
3       349892  59.0  70.0  424.0   55.0   1.7  138.0  4.5  12.6  37.0  ...   
4       353025  40.0  70.0  253.0  150.0  11.9  132.0  5.6  10.9  31.0  ...   
..         ...   ...   ...    ...    ...   ...    ...  ...   ...   ...  ...   
148     486397  20.0  70.0  123.0   44.0   1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0  60.0   97.0   18.0   1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0  60.0   95.0   46.0   0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0  80.0  100.0   50.0   1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0  60.0  105.0   49.0   1.2  150.0  4.7  15.7  44.0  ...   

     pc_n

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: 
A single column for the binary variable works even with downstream modeling; it is equivalent. On the other hand, a categorical variable that can take more than 2 values is a bit more complex. and may be more challenging for a model to handle when making a prediction.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [22]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
for column in numerical_columns:
    mean = encoded_dataset[column].mean()
    std_dev = encoded_dataset[column].std()
    upper_bound = mean + 3 * std_dev
    lower_bound = mean - 3 * std_dev
    
    encoded_dataset = encoded_dataset[(encoded_dataset[column] >= lower_bound) & (encoded_dataset[column] <= upper_bound)]


# Print the dataset
print("Dataset after removing outliers:\n")
print(encoded_dataset)

Dataset after removing outliers:

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  ...  \
2       343710  61.0   80.0  173.0  148.0  3.9  135.0  5.2   7.7  24.0  ...   
5       397388  63.0  100.0  280.0   35.0  3.2  143.0  3.5  13.0  40.0  ...   
7       438182  73.0   80.0  253.0  142.0  4.6  138.0  5.8  10.5  33.0  ...   
8       474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0  ...   
9       475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0  ...   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...  ...   
148     486397  20.0   70.0  123.0   44.0  1.0  135.0  3.8  14.6  44.0  ...   
149     203466  22.0   60.0   97.0   18.0  1.2  138.0  4.3  13.5  42.0  ...   
150     482105  44.0   60.0   95.0   46.0  0.5  138.0  4.2  15.0  50.0  ...   
151     228849  58.0   80.0  100.0   50.0  1.2  140.0  3.5  14.0  50.0  ...   
152     996076  57.0   60.0  105.0   49.0  1.2  150.0  4.7  15.7  44.0  ...   

     pc_normal  p

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [23]:
# Normalize the all Numerical Attributes in the dataset.
normalized_dataset = encoded_dataset.copy()  
normalized_dataset[numerical_columns] = (encoded_dataset[numerical_columns] - encoded_dataset[numerical_columns].min()) / \
                                         (encoded_dataset[numerical_columns].max() - encoded_dataset[numerical_columns].min())

# Print the dataset
print("Normalized dataset:\n")
print(normalized_dataset)

Normalized dataset:

     unique_id       age    bp       bgr        bu        sc       sod  \
2       343710  0.743243  0.50  0.442060  0.901961  0.432099  0.500000   
5       397388  0.770270  1.00  0.901288  0.163399  0.345679  0.766667   
7       438182  0.905405  0.50  0.785408  0.862745  0.518519  0.600000   
8       474407  0.202703  0.75  0.158798  0.196078  0.160494  0.166667   
9       475200  0.783784  1.00  0.399142  0.287582  0.839506  0.666667   
..         ...       ...   ...       ...       ...       ...       ...   
148     486397  0.189189  0.25  0.227468  0.222222  0.074074  0.500000   
149     203466  0.216216  0.00  0.115880  0.052288  0.098765  0.600000   
150     482105  0.513514  0.00  0.107296  0.235294  0.012346  0.600000   
151     228849  0.702703  0.50  0.128755  0.261438  0.098765  0.666667   
152     996076  0.689189  0.00  0.150215  0.254902  0.098765  1.000000   

          pot      hemo       pcv  ...  pc_normal  pcc_present  ba_present  \
2    0.79310

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:
One column to remove would be the "unique_id" column, as its only purpose is to provide a means of identification, which adds no value to the tasks of modeling or predictions.

In [30]:
#Remove that column
columns_to_remove = ['unique_id']
cleaned_dataset = normalized_dataset.drop(columns=columns_to_remove)

# Print the dataset
print("Dataset after removing unnecessary columns:\n")
print(cleaned_dataset)

Dataset after removing unnecessary columns:

          age    bp       bgr        bu        sc       sod       pot  \
2    0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
5    0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.206897   
7    0.905405  0.50  0.785408  0.862745  0.518519  0.600000  1.000000   
8    0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
9    0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.586207   
..        ...   ...       ...       ...       ...       ...       ...   
148  0.189189  0.25  0.227468  0.222222  0.074074  0.500000  0.310345   
149  0.216216  0.00  0.115880  0.052288  0.098765  0.600000  0.482759   
150  0.513514  0.00  0.107296  0.235294  0.012346  0.600000  0.448276   
151  0.702703  0.50  0.128755  0.261438  0.098765  0.666667  0.206897   
152  0.689189  0.00  0.150215  0.254902  0.098765  1.000000  0.620690   

         hemo       pcv      wbcc  ...  pc_normal  pcc_present  ba_present  \


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [25]:
# Export the dataframe to a new csv file
cleaned_dataset.to_csv('cleaned_dataset.csv', index=False)

# Export the dataframe to a new json file
cleaned_dataset.to_json('cleaned_dataset.json', orient='records', lines=True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [26]:
top_15_rows = cleaned_dataset.head(15)

print("Top 15 rows of the cleaned dataset:\n")
print(top_15_rows)

Top 15 rows of the cleaned dataset:

         age    bp       bgr        bu        sc       sod       pot  \
2   0.743243  0.50  0.442060  0.901961  0.432099  0.500000  0.793103   
5   0.770270  1.00  0.901288  0.163399  0.345679  0.766667  0.206897   
7   0.905405  0.50  0.785408  0.862745  0.518519  0.600000  1.000000   
8   0.202703  0.75  0.158798  0.196078  0.160494  0.166667  0.206897   
9   0.783784  1.00  0.399142  0.287582  0.839506  0.666667  0.586207   
10  0.729730  0.00  0.935622  0.169935  0.160494  0.333333  0.034483   
12  0.675676  0.75  0.253219  0.633987  0.777778  0.366667  0.655172   
13  0.851351  0.25  0.618026  0.562092  0.728395  0.000000  0.344828   
15  0.540541  0.00  0.399142  0.535948  0.358025  0.700000  0.379310   
18  0.756757  0.25  0.223176  0.209150  0.160494  0.533333  0.620690   
19  0.662162  0.50  0.618026  0.411765  0.432099  0.566667  0.689655   
20  0.783784  0.00  0.725322  0.313725  0.481481  0.566667  0.862069   
22  0.878378  0.00  0.20600

### Paste the markdown table here

| Age |    BP    |   BGR    |    BU    |    SC    |    SOD    |    POT    |    HEMO    |    PCV    |    WBCC    | RBC Normal | PC Normal | PCC Present | BA Present | HTN Yes | DM Yes | CAD Yes | Appet Poor | PE Yes | ANE Yes |
|-----|----------|----------|----------|----------|-----------|-----------|------------|-----------|------------|------------|-----------|-------------|------------|---------|--------|---------|-------------|--------|---------|
|  2  | 0.743243 | 0.50     | 0.442060 | 0.901961 | 0.432099  | 0.500000  | 0.000000   | 0.032258  | 0.395161   | False      | False     | False       | False      | True    | True   | True    | True        | True   | True    |
|  5  | 0.770270 | 1.00     | 0.901288 | 0.163399 | 0.345679  | 0.766667  | 0.524752   | 0.548387  | 0.443548   | True       | True      | False       | True       | True    | False  | True    | False       | False  | False   |
|  7  | 0.905405 | 0.50     | 0.785408 | 0.862745 | 0.518519  | 0.600000  | 0.277228   | 0.322581  | 0.233871   | False      | False     | False       | False      | True    | True   | True    | False       | False  | False   |
|  8  | 0.202703 | 0.75     | 0.158798 | 0.196078 | 0.160494  | 0.166667  | 0.059406   | 0.000000  | 0.653226   | True       | False     | True        | True       | False   | False  | False   | False       | False  | True    |
|  9  | 0.783784 | 1.00     | 0.399142 | 0.287582 | 0.839506  | 0.666667  | 0.019802   | 0.096774  | 0.258065   | False      | False     | False       | True       | True    | True   | False   | False       | True   | False   |
| 10  | 0.729730 | 0.00     | 0.935622 | 0.169935 | 0.160494  | 0.333333  | 0.019802   | 0.064516  | 0.879032   | True       | False     | True        | False      | True    | False  | False   | True        | False  | True    |
| 12  | 0.675676 | 0.75     | 0.253219 | 0.633987 | 0.777778  | 0.366667  | 0.138614   | 0.193548  | 0.169355   | False      | False     | False       | False      | True    | False  | False   | False       | False  | False   |
| 13  | 0.851351 | 0.25     | 0.618026 | 0.562092 | 0.728395  | 0.000000  | 0.168317   | 0.161290  | 0.580645   | True       | False     | True        | True       | True    | True   | True    | False       | True   | True    |
| 15  | 0.540541 | 0.00     | 0.399142 | 0.535948 | 0.358025  | 0.700000  | 0.207921   | 0.161290  | 0.830645   | True       | True      | False       | False      | True    | True   | False   | False       | False  | False   |
| 18  | 0.756757 | 0.25     | 0.223176 | 0.209150 | 0.160494  | 0.533333  | 0.485149   | 0.516129  | 0.290323   | True       | False     | False       | False      | True    | True   | False   | False       | False  | False   |
| 19  | 0.662162 | 0.50     | 0.618026 | 0.411765 | 0.432099  | 0.566667  | 0.316832   | 0.354839  | 0.250000   | True       | False     | True        | True       | True    | True   | False   | False       | False  | True    |
| 20  | 0.783784 | 0.00     | 0.725322 | 0.313725 | 0.481481  | 0.566667  | 0.178218   | 0.193548  | 0.258065   | False      | False     | False       | True       | True    | True   | False   | True        | True   | False   |
| 22  | 0.878378 | 0.00     | 0.206009 | 0.751634 | 0.604938  | 0.533333  | 0.366337   | 0.387097  | 0.879032   | True       | True      | False       | False      | True    | True   | False   | True        | True   | False   |
| 23  | 0.878378 | 0.25     | 0.639485 | 0.470588 | 0.395062  | 0.433333  | 0.267327   | 0.322581  | 0.104839   | True       | False     | True        | True       | True    | True   | True    | False       | False  | False   |
| 29  | 0.716216 | 0.50     | 1.000000 | 0.163399 | 0.111111  | 0.066667  | 0.267327   | 0.387097  | 0.532258   | False      | True      | False       | False      | False   | True   | False   | True        | False  | False   |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

In order to mitigate data conversion hallucinations from LLM conversions, it would be beneficial to feed the LLM smaller sections of the data, as opposed to the entirety of the dataset all at once. Furthemore, it would be advisable to provide clear-cut instructions that instruct the LLM how to precisely handle the data in order to avoid any clarity issues. Finally, it is important to manually verify the data conversion, informing the LLM of any issues, revising prompting, and correcting errors throughout the data conversion process.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [27]:
# Code to rename all the columns in the dataset
rename_mapping = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

cleaned_dataset.rename(columns=rename_mapping, inplace=True)

print("Dataset with renamed columns:\n")
print(cleaned_dataset)

Dataset with renamed columns:

          age  blood pressure  blood glucose random  blood urea  \
2    0.743243            0.50              0.442060    0.901961   
5    0.770270            1.00              0.901288    0.163399   
7    0.905405            0.50              0.785408    0.862745   
8    0.202703            0.75              0.158798    0.196078   
9    0.783784            1.00              0.399142    0.287582   
..        ...             ...                   ...         ...   
148  0.189189            0.25              0.227468    0.222222   
149  0.216216            0.00              0.115880    0.052288   
150  0.513514            0.00              0.107296    0.235294   
151  0.702703            0.50              0.128755    0.261438   
152  0.689189            0.00              0.150215    0.254902   

     serum creatinine    sodium  potassium  hemoglobin  packed cell volume  \
2            0.432099  0.500000   0.793103    0.000000            0.032258   
5       

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [35]:
# Converted SQL to Pandas code
result = cleaned_dataset.groupby('Target_notckd').size().reset_index(name='count')
print(result)

   Target_notckd  count
0          False     22
1           True    111


*Describe what the above code does here*

The code above finds the "Target" column (in this case, "Target_notckd"), and returns the count (number of occurrences) of a unique value in the "Target" column, which are "False" (22 occurrences) and "True" (111 occurrences).